ratis-logservice/src/main/java/org/apache/ratis/logservice/server/LogStateMachine.java

```diff
@@ -452,29 +510,194 @@ public void close() {
+    this.archiveLogName = LogServiceProtoUtil.toLogName(archiveLog.getLogName());
+    this.archiveLocation = archiveLog.getLocation();
+    this.lastArchivedIndex = archiveLog.getLastArchivedRaftIndex();
+    Throwable t = verifyState(State.ARCHIVING);
+    return CompletableFuture.completedFuture(
+        Message.valueOf(LogServiceProtoUtil.toArchiveLogReplyProto(t).toByteString()));
+  }
+
+  private CompletableFuture<Message> processArchiveLog(
+      LogServiceRequestProto logServiceRequestProto) {
+    LogServiceProtos.ArchiveLogRequestProto archiveLog = logServiceRequestProto.getArchiveLog();
+    LogName logName = LogServiceProtoUtil.toLogName(archiveLog.getLogName());
+    String loc = null;
+    this.isArchival = !archiveLog.getIsExport();
+    if(isArchival) {
+      loc = archiveLocation;
+    }else{
+      loc = archiveLog.getLocation();
+    }
+    Throwable t = null;
+    if (loc == null) {
+      t = new IllegalArgumentException(isArchival ?
+          "Location for archive is not configured" :
+          "Location for export provided is null");
+      return CompletableFuture.completedFuture(
+          Message.valueOf(LogServiceProtoUtil.toArchiveLogReplyProto(t).toByteString()));
+    }
+    final String location = loc;
+    String archiveLocationForLog = LogServiceUtils.getArchiveLocationForLog(location, logName);
+    long recordId = archiveLog.getLastArchivedRaftIndex();
+    t = verifyState(State.CLOSED);
+    if (t == null) {
+      Callable<Boolean> callable = () -> {
+        sendChangeStateRequest(State.ARCHIVING);
+        updateArchivingInfo(recordId, logName, location, isArchival);
+        ArchiveLogWriter writer = new ArchiveHdfsLogWriter();
+        writer.init(archiveLocationForLog, logName);
+        LogServiceRaftLogReader reader = new LogServiceRaftLogReader(log);
+        reader.seek(0);
+        long records = 0;
+        boolean isInterupted = false;
+        while (reader.hasNext()) {
+          writer.write(ByteBuffer.wrap(reader.next()));
+          isInterupted = Thread.currentThread().isInterrupted();
+          if (records >= DEFAULT_ARCHIVE_THRESHOLD_PER_FILE || isInterupted) {
+            commit(writer, logName, location);
+            if (isInterupted) {
+              break;
+            }
+            records = 0;
+          }
+          records++;
+        }
+        writer.close();
+        if (!isInterupted) {
+          sendChangeStateRequest(State.ARCHIVED);
+        } else {
+          try {
+            //Sleeping here before sending archival request to new leader to avoid causing problem
+            // during leader election storm
+            Thread.sleep(10000);
+          } catch (InterruptedException e) {
+          }
+          sendArchiveLogrequestToNewLeader(writer.getLastWrittenRecordId(), logName, location);
+        }
+        return true;
+      };
+      archiveFuture = executorService.submit(callable);
+    }
+    return CompletableFuture.completedFuture(
```

Comment:
Shouldn't we send back the `archiveFuture`, waiting on its completion?

Reply:
As archival is asynchronous, if we do so, woudn't the client still wait until the task completes?